unix/w0vncserver/portals/Portal.cxx

```diff
@@ -0,0 +1,252 @@
+};
+
+PortalCall::PortalCall()
+  : callback(nullptr), userData(nullptr), requestHandle(nullptr),
+    method(nullptr), signalId(0), connection(nullptr)
+{
+}
+
+PortalCall::~PortalCall()
+{
+  if (signalId > 0)
+    g_dbus_connection_signal_unsubscribe(connection, signalId);
+
+  free(requestHandle);
+}
+
+Portal::Portal() : sessionHandle_(nullptr)
+{
+  GError* error = nullptr;
+
+  connection_ = g_bus_get_sync(G_BUS_TYPE_SESSION, nullptr, &error);
+
+  if (error) {
+    std::string error_message(error->message);
+    g_error_free(error);
+    throw std::runtime_error(error_message);
+  }
+}
+
+Portal::~Portal() {
+  if (connection_) {
+    closeSession();
+    g_dbus_connection_close_sync(connection_, nullptr, nullptr);
+    g_object_unref(connection_);
+  }
+}
+
+void Portal::newRequestHandle(const char** requestHandle,
+                              const char** handleToken)
+{
+  uuid_t uuid;
+  char uuidStr[37];
+  const char* uniqueName;
+  char *cleanUniqueName;
+
+  uniqueName = g_dbus_connection_get_unique_name(connection_);
+  cleanUniqueName = (char*)malloc(strlen(uniqueName));
+
+  /* https://flatpak.github.io/xdg-desktop-portal/docs/doc-org.freedesktop.portal.Request.html#org-freedesktop-portal-request
+   * From the documentation: the caller's unique name should have its
+   * initial ':' removed, and all '.' replaced by '_'.
+   * e.g. :15759.1 -> 15759_1
+   */
+  for (uint i = 1; i < strlen(uniqueName); i++) {
+    if (uniqueName[i] == '.')
+      cleanUniqueName[i-1] = '_';
+    else
+      cleanUniqueName[i-1] = uniqueName[i];
+  }
+
+  cleanUniqueName[strlen(uniqueName)-1] = '\0';
+
+  uuid_generate(uuid);
+  uuid_unparse(uuid, uuidStr);
+
+  // Valid DBus object paths may only contain
+  // the ASCII characters "[A-Z][a-z][0-9]_"
+  for (char *p = uuidStr; *p; ++p) {
+    if (*p == '-')
+      *p = '_';
+  }
+
+  *requestHandle = strdup(core::format("/org/freedesktop/portal/desktop/request/%s/w0vncserver_%s",
+                          cleanUniqueName, uuidStr).c_str());
+
+  *handleToken = strdup(core::format("w0vncserver_%s", uuidStr).c_str());
+
+  free(cleanUniqueName);
+
+  vlog.debug("new request handle: %s", *requestHandle);
+  vlog.debug("new handle token: %s", *handleToken);
+}
+
+const char* Portal::newSessionHandle()
+{
+  const char* sessionHandle;
+  uuid_t uuid;
+  char uuidStr[37];
+
+  uuid_generate(uuid);
+  uuid_unparse(uuid, uuidStr);
+
+  // Valid DBus object paths may only contain
+  // the ASCII characters "[A-Z][a-z][0-9]_"
+  for (char *p = uuidStr; *p; ++p) {
+    if (*p == '-')
+      *p = '_';
+  }
+
+  sessionHandle = strdup(core::format("w0vncserver_%s", uuidStr).c_str());
+
+  vlog.debug("new session handle: %s", sessionHandle);
+
+  return sessionHandle;
+}
+
+void Portal::call(GDBusProxy* proxy, const char* method,
+                  GVariant* parameters, GDBusCallFlags flags,
+                  GDBusSignalCallback signalCallback,
+                  const char* requestHandle, void* userData)
```

Comment:
Given that this is now a C++ wrapper class, I would try to get rid of as much unnecessary detail as possible.

From the usage, it looks like this would suffice:

```C++
void call(GDBusProxy* proxy, const char* method, GVariant* parameters,
          GDBusCallFlags flags=G_DBUS_CALL_FLAGS_NONE, std::function<void()> callback=nullptr);
```

Avoids having the other classes having to deal with too much of the C stuff.